### PR TITLE
 **Feature:** OperationalContext hook useOperationalContext

### PR DESCRIPTION
**Pull request by adeelibr**
# Summary

Keeping the current `render props` implementation in tact, add another function called `useOperationalContext` which returns the current `context` of `OperationalContext`.

Uses Hydra pattern https://americanexpress.io/hydra/

### usage

```js
import { useOperationalContext } from "../src/OperationalContext/OperationalContext"

const Foo = () => {
  const ctx = useOperationalContext();
  return <p>Window width is {ctx.windowSize.width}</p>
};
```

P.S: I have not commited `package.json` with the updated `react` & `react-dom` version, should I for this PR?

CC: @fabien0102 @TejasQ 

# Related issue

https://github.com/contiamo/operational-ui/issues/908
https://github.com/contiamo/operational-ui/issues/910

# To be tested

Me
- [x ] No error or warning in the console on `localhost:6060`

### CODE REVIEW

**Comment by TejasQ** [comment]
> P.S: I have not commited package.json with the updated react & react-dom version, should I for this PR?

@adeelibr 💯% yes

File: src/OperationalContext/OperationalContext.init.tsx [hunk 1]
```diff
@@ -39,8 +39,12 @@ const defaultContext: Context = {
   setLoading: (_: boolean) => void 0,
 }
 
-const { Provider, Consumer: OperationalContext } = React.createContext(defaultContext)
+const OperationalContext = React.createContext(defaultContext)
 
-export default OperationalContext
+const { Provider, Consumer } = OperationalContext
+
+export const useOperationalContext = () => React.useContext(OperationalContext)
```
[the remark this answers is not in the document]
Comment (TejasQ) [reply src/OperationalContext/OperationalContext.init.tsx]: @adeelibr that's weird. Let's take a look... 👀

[the remark this answers is not in the document]
Comment (adeelibr) [reply src/OperationalContext/OperationalContext.init.tsx]: Great! regenerating lock file helped.

**Comment by TejasQ** [comment]
@adeelibr let's add an example to the README so we see it in the playground. 😉

**Comment by adeelibr** [comment]
Can you kindly review again. @TejasQ

File: src/OperationalContext/README.md [hunk 2]
```diff
@@ -14,6 +14,19 @@ Child components can get access to window dimensions as follows:
...
 
+### OperationalContext Hook
+
+You can get the value of `OperationalContext` in the shape of context.
+
+```jsx
+const { useOperationalContext } = require("./OperationalContext")
+const ILookPretty = () => {
+  const ctx = useOperationalContext()
+  return <p>{`The viewport is ${ctx.windowSize.width} pixels wide and ${ctx.windowSize.height} tall.`}</p>
```
Comment (TejasQ) [inline src/OperationalContext/README.md]:
```suggestion
  return <p>The viewport is {ctx.windowSize.width} pixels wide and {ctx.windowSize.height} tall.</p>
```

Comment (adeelibr) [reply src/OperationalContext/README.md]: I have made the changes 👍

**Comment by adeelibr** [comment]
I made one small change in `OperationalContext.tsx`

```diff
+  const OperationalContext: React.SFC<Props> = props => {
+    const ctx = useOperationalContext()
+    return <React.Fragment>{props.children({ ...ctx })}</React.Fragment>
+ }
-  const OperationalContext: React.SFC<Props> = props => (
-     <OperationalContextOriginal>{props.children}</OperationalContextOriginal>
-  );
```

This was based on https://americanexpress.io/hydra/ Which uses context under the hood, to support render-props. This way we don't have to manage both of them at the same time.

It is working fine at my end, on the local dev-server. I don't know why the deployment is failing here.

I tried running this in the `dev-server`, it works fine.

```jsx
<OperationalContext>
            {operationalContext => (
              <p>{`The viewport is ${operationalContext.windowSize.width} pixels wide and ${
                operationalContext.windowSize.height
                } tall.`}</p>
            )}
</OperationalContext>
```

**Comment by TejasQ** [comment]
@adeelibr you got it. @fabien0102 let's review this. 🚀

File: src/OperationalContext/OperationalContext.tsx [hunk 3]
```diff
@@ -16,10 +10,11 @@ export interface Props {
  * This component simply wraps OperationalContext in order to allow styleguidist to pick up on
  * it and display it in the documentation page.
  */
-const OperationalContext: React.SFC<Props> = props => (
-  <OperationalContextOriginal>{props.children}</OperationalContextOriginal>
-)
+const OperationalContext: React.SFC<Props> = props => {
+  const ctx = useOperationalContext()
+  return <React.Fragment>{props.children({ ...ctx })}</React.Fragment>
```
Comment (TejasQ) [inline src/OperationalContext/OperationalContext.tsx]: Do we need a fragment here? 🤔

Comment (adeelibr) [reply src/OperationalContext/OperationalContext.tsx]: Yes, It gives a typing error otherwise for directly returning children.

Comment (TejasQ) [reply src/OperationalContext/OperationalContext.tsx]: Have you tried 

```suggestion
  return <>{props.children({ ...ctx })}</>
```

🤔

File: dev-server/index.tsx [hunk 4]
```diff
@@ -6,7 +6,7 @@
  */
```
Comment (TejasQ) [inline dev-server/index.tsx]: ? We probably can delete this file no?

Comment (adeelibr) [reply dev-server/index.tsx]: Sorry did that by mistake, reverting it.

Comment (adeelibr) [reply dev-server/index.tsx]: Can you kindly review again. 😅 @TejasQ

**Comment by TejasQ** [comment]
@adeelibr @fabien0102 one final thing that concerns me, is that if this component is used in a class component, the hook will throw for **no fault of the user** and be a very surprising breakage iiuc. Can we protect against this somehow?

**Comment by fabien0102** [comment]
@TejasQ Can you add an example for context? I don't really when what throw 😄

**Comment by adeelibr** [comment]
The new linters rule will specify that hooks should not be used in class components. I don't see a work around for this I am afraid.

**Comment by TejasQ** [comment]
@fabien0102 @adeelibr 

here's an example

```jsx
// Legacy code

class MySomething extends React.Component {
  render() {
    <OperationalContext>
      {({ pushState }) => <a onClick={() => pushState('/my-chicken-loves-me')}>HELLO 🐔❤️</a>}
    </OperationalContext>
  }
}
```

The new `OperationalContext` will throw

```Hooks can only be used in React function components```

and the whole app will revert to an error boundary or unmount.

**Comment by adeelibr** [comment]
> The new `OperationalContext` will throw

Works fine at my end, doesn't throw an error & it won't throw an error. Because we are using that hook inside our render prop component. The way people access it using `render-props` it won't cause an error.

**Comment by TejasQ** [comment]
@adeelibr I love you let's add more hooks WOOOOOO

**Pull request merged**